### ISSUE TITLE
Transaction field `NumAccounts` returns an Invalid or unexpected value.

### ISSUE THREAD

**Issue by koen-vr** [algorand/go-algorand#3262]
## `NumAccounts` returns an Invalid or unexpected value

When trying to validate the number of accounts before accessing them I ran into some unexpected behavior. Two accounts are accessed in the accounts array/stack but `NumAccounts` tells me there is only one. There are two accounts yet when checking `NumAccounts` there is only one. In PyTeal this looks even uglier.

Teal
~~~~~
txn NumAccounts
int 1
==
....
txna Accounts 0
....
txna Accounts 1
...
~~~~~

PyTeal
~~~
Txn.accounts.length() == Int(1)
...
Txn.accounts[0]
...
Txn.accounts[1]
...
~~~

### Your environment

Install node (algod -v)
~~~
3.0.1.stable [rel/stable] (commit #b619b940)
go-algorand is licensed with AGPLv3.0
~~~
Installed with pip requirements.txt
~~~
py-algorand-sdk==1.8.0
pyteal==0.9.0
~~~

## Steps to reproduce

1 Get the demo code from repo and the scripts as instructed: 
Commit bbe77c7bd757f0dcef261cd2cd47d833ead7966e
https://github.com/vecno-io/algo-collection 

### Expected behavior

When running the setmanager.sh script I expect it to fail. The check below is incorrect as there are two accounts in the array.
https://github.com/vecno-io/algo-collection/blob/master/contracts/collection.py#L81

### Actual behavior

The check passes as the return of Txn.accounts.length() does not include the sender's address that is injected at index 0.

## Resolution

I think this should be documented and added to the spec to avoid confusion. Reading through the transaction logic code in go-algorand it is clearly defined to function and behave as is, yet the behavior falls far outside expected norms.

The other solution would be to add a +1 to offset the special case and -1 when accessing the accounts.

See:
https://github.com/algorand/go-algorand/blob/c579bf6494d6a66a9d682bc266a3b0fa46ecfe8c/data/transactions/logic/eval.go#L2018
https://github.com/algorand/go-algorand/blob/c579bf6494d6a66a9d682bc266a3b0fa46ecfe8c/data/transactions/logic/eval.go#L2015

ps. feel free to close not critical, mostly hope it saves others some headache searching while they should be working. But please do document it somewhere that is suited.

**Comment by algorandskiy** [comment]
`NumAccounts` return length of the `Accounts` field in a transaction. In your example there is only one value in the `Accounts` array and it is accessed as `txn Accounts 1` in AVM/TEAL. If a transaction has two values in the `Accounts` array, then it will be accessible as `txn Accounts 2` in AVM/TEAL. 

Zero has a special meaning in AVM/TEAL - it is always there and equals to `txn.Sender`.

We probably need to update documentation to be explicit about this.

**Comment by koen-vr** [comment]
> We probably need to update documentation to be explicit about this.

agree, it needs to be more clear. Feels awkward sins counting from zero is the standard drilled into most. Not counting the sender as being part of the array is a missed UX opportunity that good documentation has to fix.
